### ISSUE TITLE
Loki exhausts ephemeral ports when loading labels

### ISSUE THREAD

**Issue by JensErat** [grafana/loki#1938]
**Describe the bug**

When listing labels (opening explore view in Grafana), querier seems to make an individual S3 call for each stream. This exhausted the ephemeral ports in its container. To makes things worse, this also blocked storing them in memcached...

Our solution was to limit parallel connections to S3, see https://github.com/weaveworks/common/pull/186 where I proposed my change. This might also be possible in Loki/Cortex by limiting in-flight S3 requests, but I guess delegating that to the golang HTTP client is more reasonable (they have a queue already). But this depends on what weaveworks says.

**To Reproduce**
1. Started Loki v1.3.0
2. Push a more or less large number of streams per tenant to Loki
3. Open explore view for Loki tenant in Grafana

**Expected behavior**

Labels load, I can query Loki

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: own helm chart

**Screenshots, Promtail config, or terminal output**

- `netstat -pan`/`ss -pan` lists _lots_ of connections to S3 backend in `TIME_WAIT`
- timeout error messages towards S3/memcached

**Comment by slim-bean** [comment]
Which schema are you using currently?  The v11 schema handles label queries much more efficiently as the label names are stored in the index to avoid having to download chunks.

**Comment by JensErat** [comment]
I see we're still using v10. I will research a little bit regarding the schema. I guess we'll have to drop old data (would be fine, we're just starting to use Loki in production)?

**Comment by cyriltovena** [comment]
No you can add a new schema and keep your data.

Let;s say you have this:

```
  schema_config:
    configs:
    - from: 2018-04-15
      store: boltdb
      object_store: filesystem
      schema: v9
      index:
        prefix: index_
        period: 168h
```

change it to this

```
  schema_config:
    configs:
    - from: 2018-04-15
      store: boltdb
      object_store: filesystem
      schema: v9
      index:
        prefix: index_
        period: 168h
    - from: 2019-04-15
      store: boltdb
      object_store: filesystem
      schema: v11
      index:
        prefix: index_v11_
        period: 168h
```

And you just have to wait the `2019-04-15` it will automatically swift to the new format, and the old format can still be queried normaly.

**Comment by slim-bean** [comment]
It's not necessary to change the prefix of the index, it's fine to use the same prefix for multiple schemas.

Also be aware that the  label query performance will only be seen for label queries using the new schema, as Cyril's stated though, all log data will still be queryable

**Comment by JensErat** [comment]
I can hardly find reference regarding the schema version. Will I have to upgrade to Loki 1.4+ (without losing old data, this will require me to rebase my patch and recompile) or can I use schema v11 with Loki 1.3? I guess if I keep the old data with schema v10 and upgrade I'll have the exhaustion issue again. Dropping the old data might still be an alternative for us in this case, though. I guess we'll have to do so in at least one of our installations, so we can decide whether v11 _without_ the limit actually solves the issue.

I guess some kind of parallel connection limit might still be a good thing to do. But I'll have a try  with v11 tomorrow.

**Comment by slim-bean** [comment]
The documentation around schema is very inadequate currently, our apologies for that.  There is an issue to fix this but it hasn't been done yet #1851 .

The v11 schema will work just fine with Loki 1.3
